src/audio/drc/drc_generic.c

```diff
@@ -0,0 +1,469 @@
+		 * ratio portion is smooth (1st derivative matched).
+		 */
+		gain = volume_gain(p, abs_input_array[i]);
+		is_release = (gain > detector_average);
+		if (is_release) {
+			if (gain > DRC_NEG_TWO_DB) {
+				detector_average +=
+					(gain - detector_average) *
+					sat_release_rate_at_neg_two_db;
+			} else {
+				db_per_frame =
+					linear_to_decibels(gain) * sat_release_frames_inv_neg;
+				sat_release_rate =
+					decibels_to_linear(db_per_frame) - 1;
+				detector_average += (gain - detector_average) *
+						    sat_release_rate;
+			}
+		} else {
+			detector_average = gain;
+		}
+
+		/* Fix gremlins. */
+		if (isbadf(detector_average))
+			detector_average = 1.0f;
+		else
+			detector_average = MIN(detector_average, 1.0f);
+	}
+
+	state->detector_average = Q_CONVERT_FLOAT(detector_average, 30);
+}
+
+/* Updates the envelope_rate used for the next division */
+static void drc_update_envelope(struct drc_state *state, const struct sof_drc_params *p)
+{
+	const float kA = Q_CONVERT_QTOF(p->kA, 12);
+	const float kB = Q_CONVERT_QTOF(p->kB, 12);
+	const float kC = Q_CONVERT_QTOF(p->kC, 12);
+	const float kD = Q_CONVERT_QTOF(p->kD, 12);
+	const float kE = Q_CONVERT_QTOF(p->kE, 12);
+	const float attack_frames = Q_CONVERT_QTOF(p->attack_frames, 20);
+
+	/* Calculate desired gain */
+	float desired_gain = Q_CONVERT_QTOF(state->detector_average, 30);
+
+	/* Pre-warp so we get desired_gain after sin() warp below. */
+	float scaled_desired_gain = warp_asinf(desired_gain);
+
+	/* Deal with envelopes */
+
+	/* envelope_rate is the rate we slew from current compressor level to
+	 * the desired level.  The exact rate depends on if we're attacking or
+	 * releasing and by how much.
+	 */
+	float envelope_rate;
+
+	int is_releasing = scaled_desired_gain > Q_CONVERT_QTOF(state->compressor_gain, 30);
+
+	/* compression_diff_db is the difference between current compression
+	 * level and the desired level. */
+	float compression_diff_db = linear_to_decibels(
+		Q_CONVERT_QTOF(state->compressor_gain, 30) / scaled_desired_gain);
+
+	float x, x2, x3, x4;
+	float release_frames;
+	float db_per_frame;
+	float eff_atten_diff_db;
+
+	if (is_releasing) {
+		/* Release mode - compression_diff_db should be negative dB */
+		state->max_attack_compression_diff_db = INT32_MIN;
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
+			compression_diff_db = -1;
+
+		/* Adaptive release - higher compression (lower
+		 * compression_diff_db) releases faster. Contain within range:
+		 * -12 -> 0 then scale to go from 0 -> 3
+		 */
+		x = compression_diff_db;
+		x = MAX(-12.0f, x);
+		x = MIN(0.0f, x);
+		x = 0.25f * (x + 12);
+
+		/* Compute adaptive release curve using 4th order polynomial.
+		 * Normal values for the polynomial coefficients would create a
+		 * monotonically increasing function.
+		 */
+		x2 = x * x;
+		x3 = x2 * x;
+		x4 = x2 * x2;
+		release_frames = kA + kB * x + kC * x2 + kD * x3 + kE * x4;
+
+#define kSpacingDb 5
```

Comment:
we should add these config fields to the Kconfig and adjust their names as needed